File: app/src/main/java/com/phlox/tvwebbrowser/activity/singleton/shortcuts/ShortcutMgr.java
package com.phlox.tvwebbrowser.activity.singleton.shortcuts;

import android.content.Context;
import android.content.SharedPreferences;
import android.view.KeyEvent;

import com.phlox.tvwebbrowser.activity.main.MainActivity;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by PDT on 06.08.2017.
 */

public class ShortcutMgr {
    public static final String PREFS_SHORTCUTS = "shortcuts";
    private static ShortcutMgr instance;
    private final Context ctx;
    private Map<Integer, Shortcut> shortcuts;

    public static ShortcutMgr getInstance(Context context) {
        if (instance == null) {
            instance = new ShortcutMgr(context.getApplicationContext());
        }
        return instance;
    }

    private ShortcutMgr(Context applicationContext) {
        this.ctx = applicationContext;
        shortcuts = new HashMap<>();
        SharedPreferences prefs = ctx.getSharedPreferences(PREFS_SHORTCUTS, Context.MODE_PRIVATE);
        for (Shortcut shortcut: Shortcut.values()) {
            shortcut.keyCode = prefs.getInt(shortcut.prefsKey, shortcut.keyCode);
            if (shortcut.keyCode != 0) {
                shortcuts.put(shortcut.keyCode, shortcut);
            }
        }
    }

    public void save(Shortcut shortcut) {
        int oldKey = 0;
        for (Map.Entry<Integer, Shortcut> entry: shortcuts.entrySet()) {
            if (entry.getValue().equals(shortcut)) {
                oldKey = entry.getKey();
            }
        }
        if (oldKey != 0) {
            shortcuts.remove(oldKey);
        }
        if (shortcut.keyCode != 0) {
            shortcuts.put(shortcut.keyCode, shortcut);
        }
        SharedPreferences prefs = ctx.getSharedPreferences(PREFS_SHORTCUTS, Context.MODE_PRIVATE);
        prefs.edit()
                .putInt(shortcut.prefsKey, shortcut.keyCode)
                .apply();
    }

    public Shortcut findForMenu(int menuId) {
        for (Map.Entry<Integer, Shortcut> entry: shortcuts.entrySet()) {
            if (entry.getValue().menuItemId == menuId) {
                return entry.getValue();
            }
        }
        return Shortcut.findForMenu(menuId);
    }

    public boolean process(int keyCode, MainActivity mainActivity) {
        Shortcut shortcut = shortcuts.get(keyCode);
        if (shortcut == null) {
            return false;
        }
        switch (shortcut) {
            case MENU:
                mainActivity.toggleMenu();
                return true;
            case NAVIGATE_BACK:
                mainActivity.navigateBack();
                return true;
            case NAVIGATE_HOME:
                mainActivity.navigate(MainActivity.HOME_URL);
                return true;
            case REFRESH_PAGE:
                mainActivity.refresh();
                return true;
            case VOICE_SEARCH:
                mainActivity.initiateVoiceSearch();
                return true;
        }
        return false;
    }

    public boolean canProcessKeyCode(int keyCode) {
        return shortcuts.get(keyCode) != null;
    }
}

File: app/src/main/java/com/phlox/tvwebbrowser/singleton/shortcuts/ShortcutMgr.kt
package com.phlox.tvwebbrowser.singleton.shortcuts

import android.content.Context
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import android.view.KeyEvent
import androidx.annotation.UiThread
import com.phlox.tvwebbrowser.TVBro
import com.phlox.tvwebbrowser.Config

import com.phlox.tvwebbrowser.activity.main.MainActivity
import com.phlox.tvwebbrowser.model.WebTabState
import com.phlox.tvwebbrowser.webengine.WebEngine

/**
 * Created by PDT on 06.08.2017.
 */

class ShortcutMgr private constructor() {
    private val shortcuts = ArrayList<Shortcut>()
    private var trackingShortcuts: List<Shortcut>? = null
    private val prefs: SharedPreferences =
        TVBro.instance.getSharedPreferences(PREFS_SHORTCUTS, Context.MODE_PRIVATE)
    private val uiHandler = Handler(Looper.getMainLooper())

    init {
        for (shortcut in Shortcut.entries) {
            shortcut.keyCode = prefs.getInt(shortcut.prefsKey, shortcut.keyCode)
            shortcut.modifiers = prefs.getInt(shortcut.prefsKey + "_mod", shortcut.modifiers)
            shortcut.longPressFlag = prefs.getBoolean(shortcut.prefsKey + "_lp", shortcut.longPressFlag)
            if (shortcut.keyCode != 0) {
                shortcuts.add(shortcut)
            }
        }
    }

    fun save(shortcut: Shortcut) {
        if (shortcut.keyCode == 0) {
            prefs.edit()
                    .remove(shortcut.prefsKey)
                    .remove(shortcut.prefsKey + "_mod")
                    .remove(shortcut.prefsKey + "_lp")
                    .apply()
            shortcuts.remove(shortcut)
            return
        }
        if (!shortcuts.contains(shortcut)) {
            shortcuts.add(shortcut)
        }
        prefs.edit()
                .putInt(shortcut.prefsKey, shortcut.keyCode)
                .putInt(shortcut.prefsKey + "_mod", shortcut.modifiers)
                .putBoolean(shortcut.prefsKey + "_lp", shortcut.longPressFlag)
                .apply()
    }

    fun findForId(id: Int): Shortcut {
        val shortcut = Shortcut.entries[id]
        for (s in shortcuts) {
            if (s == shortcut) {
                return s
            }
        }
        return shortcut
    }

    @UiThread
    fun process(shortcut: Shortcut, mainActivity: MainActivity, webEngine: WebEngine?) {
        when (shortcut) {
            Shortcut.MENU -> {
                mainActivity.toggleMenu()
            }
            Shortcut.NAVIGATE_BACK -> {
                mainActivity.navigateBack()
            }
            Shortcut.NAVIGATE_HOME -> {
                mainActivity.navigate(Config.HOME_URL_ALIAS)
            }
            Shortcut.REFRESH_PAGE -> {
                mainActivity.refresh()
            }
            Shortcut.VOICE_SEARCH -> {
                mainActivity.initiateVoiceSearch()
            }
            Shortcut.PLAY_PAUSE -> {
                webEngine?.togglePlayback()
            }
            Shortcut.MEDIA_STOP -> {
                webEngine?.stopPlayback()
            }
            Shortcut.MEDIA_REWIND -> {
                webEngine?.rewind()
            }
            Shortcut.MEDIA_FAST_FORWARD -> {
                webEngine?.fastForward()
            }
        }
    }

    private fun shortCutsForEvent(keyCode: Int, modifiers: Int): List<Shortcut> {
        val findings = ArrayList<Shortcut>()
        for (shortcut in shortcuts) {
            if (shortcut.keyCode == keyCode) {
                if (shortcut.modifiers == modifiers) {
                    findings.add(shortcut)
                }
            }
        }
        return findings
    }

    private fun onKeyDown(event: KeyEvent, mainActivity: MainActivity, tab: WebTabState?): Boolean {
        val shortcuts = shortCutsForEvent(event.keyCode, event.modifiers)
        if (shortcuts.isEmpty()) return false
        trackingShortcuts = shortcuts
        if (event.repeatCount == 0) {
            event.startTracking()
        }
        if (event.isLongPress) {
            return onKeyLongPress(event, mainActivity, tab)
        }
        return true
    }

    private fun onKeyUp(event: KeyEvent, mainActivity: MainActivity, tab: WebTabState?): Boolean {
        val trackingShortcuts = trackingShortcuts ?: return false
        for (shortcut in trackingShortcuts) {
            if (shortcut.longPressFlag || event.modifiers != shortcut.modifiers) {
                continue
            }
            uiHandler.post { process(shortcut, mainActivity, tab?.webEngine) }
            this.trackingShortcuts = null
            return true
        }
        return false
    }

    private fun onKeyLongPress(event: KeyEvent, mainActivity: MainActivity, tab: WebTabState?): Boolean {
        val trackingShortcuts = trackingShortcuts ?: return false
        for (shortcut in trackingShortcuts) {
            if (!shortcut.longPressFlag || event.modifiers != shortcut.modifiers) {
                continue
            }
            uiHandler.post { process(shortcut, mainActivity, tab?.webEngine) }
            this.trackingShortcuts = null
            return true
        }
        return false
    }

    fun handle(event: KeyEvent, mainActivity: MainActivity, value: WebTabState?): Boolean {
        return when (event.action) {
            KeyEvent.ACTION_DOWN -> onKeyDown(event, mainActivity, value)
            KeyEvent.ACTION_UP -> onKeyUp(event, mainActivity, value)
            else -> false
        }
    }

    fun tryHandleEmulatedSimpleKeyPress(keyCode: Int, mainActivity: MainActivity, tab: WebTabState?): Boolean {
        val shortcuts = shortCutsForEvent(keyCode, 0)
        if (shortcuts.isEmpty()) return false
        uiHandler.post { process(shortcuts.first(), mainActivity, tab?.webEngine) }
        return true
    }

    companion object {
        const val PREFS_SHORTCUTS = "shortcuts"

        private var instance: ShortcutMgr? = null

        @Synchronized fun getInstance(): ShortcutMgr {
            if (instance == null) {
                instance = ShortcutMgr()
            }
            return instance!!
        }
    }
}
